rust/datafusion/src/logicalplan.rs

```diff
@@ -363,8 +335,6 @@ pub enum Expr {
         name: String,
         /// List of expressions to feed to the functions as arguments
         args: Vec<Expr>,
-        /// The `DataType` the expression will yield
-        return_type: DataType,
```

Comment:
I think (eventually) `AggregateFunction` store its return type will be important for user defined aggregates, but for now hard coding the rules to compute the output type based on the input type for built in aggregates is a good idea.

Perhaps we can special case UDAs in a follow on PR with something such as `UserDefinedAggregateFunction` rather than making `AggregateFunction` work for both built in aggregates and user defined ones

Reply:
I agree with this: we need two interfaces: one internal that supports variable types, so that we abstract out the hard-coded return types and make them functions of the expression, and a public one for users to declared fixed-type UDFs.